File: backend/app/schemas/analytics.py
from datetime import date as date_type
from enum import StrEnum

from pydantic import BaseModel


class Confidence(StrEnum):
    """How much weight a reader should put on a derived number.

    Present on every block that extrapolates rather than counts. §16 forbids
    fake precision, and the honest way to keep a thin-data estimate useful is
    to ship it labelled rather than to ship it bare.
    """

    LOW = "low"
    MODERATE = "moderate"
    HIGH = "high"


class TrendDirection(StrEnum):
    FALLING = "falling"
    STEADY = "steady"
    RISING = "rising"


class WeightTrendPoint(BaseModel):
    date: date_type
    #: The day's logged weight (mean of that day's entries). Never interpolated
    #: — a day without a weigh-in produces no point at all rather than a made-up
    #: one, so the chart cannot imply data that was never recorded.
    weight_kg: float
    #: Smoothed trend value at this point.
    trend_kg: float


class GoalProjection(BaseModel):
    target_weight_kg: float
    estimated_weeks: float
    estimated_date: date_type
    caveat: str


class WeightTrendOut(BaseModel):
    available: bool
    #: Set when `available` is false: why there is no trend, in plain language.
    unavailable_reason: str | None = None
    points: list[WeightTrendPoint] = []
    entry_days: int = 0
    span_days: int = 0
    #: Rate of change from a least-squares fit, not from first-minus-last:
    #: two noisy endpoints can disagree wildly with the line through everything
    #: between them. None when the fit isn't defined.
    weekly_change_kg: float | None = None
    direction: TrendDirection | None = None
    #: Fraction of the weight variance the straight line explains (0–1).
    fit_quality: float | None = None
    confidence: Confidence | None = None
    #: Absent unless there is a goal weight *and* the trend is moving toward it
    #: at a rate worth extrapolating.
    goal_projection: GoalProjection | None = None
    projection_unavailable_reason: str | None = None


class TargetAdherence(BaseModel):
    """Days on target out of days that could be judged.

    Both numbers are returned rather than a percentage: "4 of 5 logged days" is
    a very different claim from "4 of 30 days", and a bare 80% hides which one
    it is.
    """

    days_on_target: int
    days_counted: int


class AdherenceOut(BaseModel):
    days_in_window: int
    days_food_logged: int
    current_streak_days: int
    longest_streak_days: int
    calories: TargetAdherence
    protein: TargetAdherence
    water: TargetAdherence
    #: False when there's no active goal — the three blocks above are then all
    #: zero-of-zero, because there is no target to be on.
    has_active_goal: bool


class MacroSplitOut(BaseModel):
    available: bool
    unavailable_reason: str | None = None
    days_counted: int = 0
    protein_percent: float | None = None
    carbs_percent: float | None = None
    fat_percent: float | None = None
    target_protein_percent: float | None = None
    target_carbs_percent: float | None = None
    target_fat_percent: float | None = None


class WeekdayPattern(BaseModel):
    #: 0 = Monday, matching `date.weekday()`.
    weekday: int
    label: str
    days_sampled: int
    #: None when nothing was logged on that weekday in the window. Zero would
    #: claim the user ate nothing; None says we don't know.
    avg_calories_kcal: float | None
    avg_activity_minutes: float | None


class EnergyBalanceOut(BaseModel):
    available: bool
    unavailable_reason: str | None = None
    days_counted: int = 0
    avg_intake_kcal: float | None = None
    #: Estimated from the profile (Mifflin-St Jeor × activity multiplier), the
    #: same deterministic path goal targets use.
    estimated_expenditure_kcal: int | None = None
    net_kcal: float | None = None
    #: Burn as reported by the user or their device, shown for context and
    #: deliberately *not* added to the estimate above — the activity multiplier
    #: already accounts for exercise, so summing the two would double-count it.
    reported_activity_burn_kcal: int | None = None
    note: str | None = None


class AnalyticsSummaryOut(BaseModel):
    generated_for: date_type
    since: date_type
    window_days: int
    weight_trend: WeightTrendOut
    adherence: AdherenceOut
    macro_split: MacroSplitOut
    weekday_patterns: list[WeekdayPattern]
    energy_balance: EnergyBalanceOut


class AdaptiveTargetBasis(BaseModel):
    """The evidence behind an adaptive suggestion, returned alongside it.

    A number that changes someone's daily calorie target should be auditable
    by the person it is aimed at, so the inputs travel with the output rather
    than staying on the server.
    """

    window_days: int
    span_days: int
    days_with_food_logged: int
    weigh_in_days: int
    avg_intake_kcal: float
    weight_change_kg: float


class AdaptiveTargetsOut(BaseModel):
    available: bool
    unavailable_reason: str | None = None
    confidence: Confidence | None = None
    #: Maintenance implied by the user's own intake and weight change.
    estimated_maintenance_kcal: int | None = None
    #: Maintenance predicted from the profile (Mifflin-St Jeor x activity
    #: level), shown next to the observed figure so the difference between a
    #: population equation and this person is visible rather than hidden.
    predicted_maintenance_kcal: int | None = None
    current_target_calories: int | None = None
    suggested_target_calories: int | None = None
    suggested_protein_g: int | None = None
    suggested_carbs_g: int | None = None
    suggested_fat_g: int | None = None
    delta_kcal: int | None = None
    basis: AdaptiveTargetBasis | None = None
    #: Always populated when a suggestion is present — including the standing
    #: reminder that nothing has been applied.
    caveats: list[str] = []

File: mobile/src/api/analytics.ts
import { apiRequest } from "./client";

export type Confidence = "low" | "moderate" | "high";
export type TrendDirection = "falling" | "steady" | "rising";

export interface WeightTrendPoint {
  date: string;
  /** The day's logged weight. Never interpolated — days without a weigh-in produce no point. */
  weight_kg: number;
  trend_kg: number;
}

export interface GoalProjection {
  target_weight_kg: number;
  estimated_weeks: number;
  estimated_date: string;
  caveat: string;
}

export interface WeightTrendOut {
  available: boolean;
  unavailable_reason: string | null;
  points: WeightTrendPoint[];
  entry_days: number;
  span_days: number;
  /** null when there aren't enough weigh-ins, over enough time, to fit a rate. */
  weekly_change_kg: number | null;
  direction: TrendDirection | null;
  fit_quality: number | null;
  confidence: Confidence | null;
  goal_projection: GoalProjection | null;
  projection_unavailable_reason: string | null;
}

/** Days on target out of days that could be judged — both numbers, never a bare percentage. */
export interface TargetAdherence {
  days_on_target: number;
  days_counted: number;
}

export interface AdherenceOut {
  days_in_window: number;
  days_food_logged: number;
  current_streak_days: number;
  longest_streak_days: number;
  calories: TargetAdherence;
  protein: TargetAdherence;
  water: TargetAdherence;
  has_active_goal: boolean;
}

export interface MacroSplitOut {
  available: boolean;
  unavailable_reason: string | null;
  days_counted: number;
  protein_percent: number | null;
  carbs_percent: number | null;
  fat_percent: number | null;
  target_protein_percent: number | null;
  target_carbs_percent: number | null;
  target_fat_percent: number | null;
}

export interface WeekdayPattern {
  /** 0 = Monday. */
  weekday: number;
  label: string;
  days_sampled: number;
  /** null when nothing was logged on that weekday — not 0, which would claim the user ate nothing. */
  avg_calories_kcal: number | null;
  avg_activity_minutes: number | null;
}

export interface EnergyBalanceOut {
  available: boolean;
  unavailable_reason: string | null;
  days_counted: number;
  avg_intake_kcal: number | null;
  estimated_expenditure_kcal: number | null;
  net_kcal: number | null;
  /** Shown for context and deliberately not added to expenditure — see `note`. */
  reported_activity_burn_kcal: number | null;
  note: string | null;
}

export interface AnalyticsSummaryOut {
  generated_for: string;
  since: string;
  window_days: number;
  weight_trend: WeightTrendOut;
  adherence: AdherenceOut;
  macro_split: MacroSplitOut;
  weekday_patterns: WeekdayPattern[];
  energy_balance: EnergyBalanceOut;
}

export interface AdaptiveTargetBasis {
  window_days: number;
  span_days: number;
  days_with_food_logged: number;
  weigh_in_days: number;
  avg_intake_kcal: number;
  weight_change_kg: number;
}

export interface AdaptiveTargetsOut {
  available: boolean;
  unavailable_reason: string | null;
  confidence: Confidence | null;
  estimated_maintenance_kcal: number | null;
  predicted_maintenance_kcal: number | null;
  current_target_calories: number | null;
  suggested_target_calories: number | null;
  suggested_protein_g: number | null;
  suggested_carbs_g: number | null;
  suggested_fat_g: number | null;
  delta_kcal: number | null;
  basis: AdaptiveTargetBasis | null;
  caveats: string[];
}

export const analyticsApi = {
  summary: (days?: number): Promise<AnalyticsSummaryOut> =>
    apiRequest<AnalyticsSummaryOut>(
      `/api/v1/analytics/summary${days ? `?days=${days}` : ""}`,
    ),

  /**
   * A calorie target proposed from the user's own intake and weight history.
   * Advisory only — the server never applies it, and neither does this client.
   */
  adaptiveTargets: (): Promise<AdaptiveTargetsOut> =>
    apiRequest<AdaptiveTargetsOut>("/api/v1/analytics/adaptive-targets"),
};
